### ISSUE TITLE
Python: Additional channel metadata / Proxy auth digest

### ISSUE THREAD

**Issue by tkrause** [grpc/grpc#30898]
### Is your feature request related to a problem? Please describe.
We have a Python gRPC client that is used to communicate with some of our backend gRPC services. However, one of our large customers has a corporate proxy that requires `digest authentication`. 

Currently the Python client does not support any of the following which could allow this to work:

- Metadata at the channel level (i.e sending headers on the connect request)
- Ability to implement ProxyMapper 
- Ability to implement socket mutator

### Describe the solution you'd like
Ideally, a developer should be able to specify `metadata` for a gRPC channel in the options when it is instantiated.

This metadata would then be sent with the initial gRPC connection request but not on calls to stubs.

### Describe alternatives you've considered
- Proxy Mapper (not supported in Python)
- Socket Mutator (not supported in Python)
- Interceptors (only works for calls to stubs)

### Additional context
Long story short, we need to be able to set a header (metadata) on the initial gRPC request to pass through the proxy. The value of the token will change for every new channel.

We've provided the customer an abstraction layer to be able to set this token which works for our HTTP/HTTPS clients. But we have no way of being able to set this in the Python gRPC client.

**Comment by tkrause** [comment]
After some digging into the C++ source, we've come up with a work around in the meantime.

However, this does not seem to be an officially supported option as it is not mentioned anywhere. It only seems to be referenced in [http_connect_handshaker.h](https://github.com/grpc/grpc/blob/063c36cb46733c13d2ce8116b6af482c9bd832d6/src/core/lib/transport/http_connect_handshaker.h#L33), [http_connect_handshaker.cc](https://github.com/grpc/grpc/blob/d304712f649da81dc403b1f9f531e4c2bbe4fde8/src/core/lib/transport/http_connect_handshaker.cc#L313) and [http_proxy.cc](https://github.com/grpc/grpc/blob/d304712f649da81dc403b1f9f531e4c2bbe4fde8/src/core/ext/filters/client_channel/http_proxy.cc#L196).

If it is to be officially supported, it should probably be documented as a `ChannelOption` or in https://github.com/grpc/grpc/blob/v1.46.x/include/grpc/impl/codegen/grpc_types.h which is referenced in the [Python docs](https://grpc.github.io/grpc/python/grpc.html).

Work around:

```
options = (
    # you can add as many headers as you want, each one needs to be formatted as valid
    # key: value pairs.

    # additional headers can be added with a newline
    ("grpc.http_connect_headers", "x-test-header: hello\nx-test-header2: hello again"),
)

host, port = "some-host", 443
chan = grpc.secure_channel("{}:{}".format(host, port), credentials, options)
```

**Comment by gnossen** [comment]
@markdroth Thoughts on the above?

**Comment by markdroth** [comment]
I would not be opposed to exposing the `GRPC_ARG_HTTP_CONNECT_SERVER` and `GRPC_ARG_HTTP_CONNECT_HEADERS` channel args.  Note that this would be a bit of a blunt hammer in that it would set the behavior for all connections for a given channel, rather than being able to set it dynamically based on the address; the latter could be done via a proxy mapper, but that's not a public API, and I'd rather not make it one.

I don't think socket mutators have anything to do with this.  Even if we exposed that API, I don't think it would solve this problem.

@yashykt owns the HTTP CONNECT code, so he's probably the right person to address this.
